instrumentation/runtime-metrics/library/src/main/java/io/opentelemetry/instrumentation/runtimemetrics/Threads.java

```diff
@@ -49,14 +61,58 @@ void registerObservers(OpenTelemetry openTelemetry, ThreadMXBean threadBean) {
         .setDescription("Number of executing threads")
         .setUnit("1")
         .buildWithCallback(
-            observableMeasurement -> {
-              observableMeasurement.record(
-                  threadBean.getDaemonThreadCount(),
-                  Attributes.builder().put(DAEMON, true).build());
-              observableMeasurement.record(
-                  threadBean.getThreadCount() - threadBean.getDaemonThreadCount(),
-                  Attributes.builder().put(DAEMON, false).build());
-            });
+            isJava9OrNewer() ? java9AndNewerCallback(threadBean) : java8Callback(threadBean));
+  }
+
+  @Nullable private static final MethodHandle THREAD_INFO_IS_DAEMON;
+
+  static {
+    MethodHandle isDaemon;
+    try {
+      isDaemon =
+          MethodHandles.publicLookup()
+              .findVirtual(ThreadInfo.class, "isDaemon", MethodType.methodType(boolean.class));
+    } catch (NoSuchMethodException | IllegalAccessException e) {
+      isDaemon = null;
+    }
+    THREAD_INFO_IS_DAEMON = isDaemon;
+  }
+
+  private static boolean isJava9OrNewer() {
+    return THREAD_INFO_IS_DAEMON != null;
+  }
+
+  private static Consumer<ObservableLongMeasurement> java8Callback(ThreadMXBean threadBean) {
+    return measurement -> {
+      measurement.record(
+          threadBean.getDaemonThreadCount(), Attributes.builder().put(DAEMON, true).build());
+      measurement.record(
+          threadBean.getThreadCount() - threadBean.getDaemonThreadCount(),
+          Attributes.builder().put(DAEMON, false).build());
+    };
+  }
+
+  private static Consumer<ObservableLongMeasurement> java9AndNewerCallback(
+      ThreadMXBean threadBean) {
+    return measurement -> {
+      Map<Attributes, Long> counts = new HashMap<>();
+      long[] threadIds = threadBean.getAllThreadIds();
+      for (ThreadInfo threadInfo : threadBean.getThreadInfo(threadIds)) {
+        Attributes threadAttributes = threadAttributes(threadInfo);
```

Comment:
(i've encountered this in the wild)
```suggestion
        if (threadInfo == null) {
          continue;
        }
        Attributes threadAttributes = threadAttributes(threadInfo);
```